src/tcpclientsrc.rs

```diff
@@ -0,0 +1,762 @@
+                let mut settings = self.settings.lock().unwrap();
+                settings.chunk_size = value.get().unwrap();
+            }
+            Property::String("context", ..) => {
+                let mut settings = self.settings.lock().unwrap();
+                settings.context = value.get().unwrap_or_else(|| "".into());
+            }
+            Property::Int("context-threads", ..) => {
+                let mut settings = self.settings.lock().unwrap();
+                settings.context_threads = value.get().unwrap();
+            }
+            Property::UInt("context-wait", ..) => {
+                let mut settings = self.settings.lock().unwrap();
+                settings.context_wait = value.get().unwrap();
+            }
+            _ => unimplemented!(),
+        }
+    }
+
+    fn get_property(&self, _obj: &glib::Object, id: u32) -> Result<glib::Value, ()> {
+        let prop = &PROPERTIES[id as usize];
+
+        match *prop {
+            Property::String("address", ..) => {
+                let mut settings = self.settings.lock().unwrap();
+                Ok(settings.address.to_value())
+            }
+            Property::UInt("port", ..) => {
+                let mut settings = self.settings.lock().unwrap();
+                Ok(settings.port.to_value())
+            }
+            Property::Boxed("caps", ..) => {
+                let mut settings = self.settings.lock().unwrap();
+                Ok(settings.caps.to_value())
+            }
+            Property::UInt("chunk-size", ..) => {
+                let mut settings = self.settings.lock().unwrap();
+                Ok(settings.chunk_size.to_value())
+            }
+            Property::String("context", ..) => {
+                let mut settings = self.settings.lock().unwrap();
+                Ok(settings.context.to_value())
+            }
+            Property::Int("context-threads", ..) => {
+                let mut settings = self.settings.lock().unwrap();
+                Ok(settings.context_threads.to_value())
+            }
+            Property::UInt("context-wait", ..) => {
+                let mut settings = self.settings.lock().unwrap();
+                Ok(settings.context_wait.to_value())
+            }
+            _ => unimplemented!(),
+        }
+    }
+}
+
+impl ElementImpl<Element> for TcpClientSrc {
+    fn change_state(
+        &self,
+        element: &Element,
+        transition: gst::StateChange,
+    ) -> gst::StateChangeReturn {
+        gst_trace!(self.cat, obj: element, "Changing state {:?}", transition);
+
+        match transition {
+            gst::StateChange::NullToReady => match self.prepare(element) {
+                Err(err) => {
+                    element.post_error_message(&err);
+                    return gst::StateChangeReturn::Failure;
+                }
+                Ok(_) => (),
+            },
+            gst::StateChange::PlayingToPaused => match self.stop(element) {
```

Comment:
As this is not a live source, reading should be started/stopped when switching between `Ready` and `Paused`. There should be no difference between `Paused` and `Playing`

Reply:
@sdroege This is the last one that needs to be corrected.

I understand your comment, but I'm not sure how to change it.
Is it correct to call push_buffer function in Paused state?

Reply:
It's correct to push buffers in Paused state, yes

Reply:
Thank you, I submit new commit to fix it. pls review it